### ISSUE TITLE
Issue with using Declarative API with custom (non-dataclass) classes (from 1.4.11)

### ISSUE THREAD

**Issue by subhashb** [sqlalchemy/sqlalchemy#6791]
**Describe the bug**
Instantiating a custom model class with the Declarative API does not work after changes that addressed #6331

I am dynamically creating and instantiating a Model class whose attributes are derived from a custom data container (not python `dataclasses`). This works for SQLAlchemy <=1.4.10 but breaks with changes that addressed #6331 as part of version 1.4.11.

**To Reproduce**

MVCE:

```py
import re

from abc import ABCMeta, abstractmethod

from sqlalchemy import Column
from sqlalchemy.ext import declarative as sa_dec
from sqlalchemy import types as sa_types
from sqlalchemy.ext.declarative import as_declarative, declared_attr

sample_definition = {
    "id": int,
    "name": str,
    "age": float,
    "adult": bool,
}


class DeclarativeMeta(sa_dec.DeclarativeMeta, ABCMeta):
    """ Metaclass for the Sqlalchemy declarative schema """

    def __init__(cls, classname, bases, dict_):
        # Update the class attrs with the entity attributes

        field_mapping = {
            bool: sa_types.Boolean,
            dict: sa_types.PickleType,
            float: sa_types.Float,
            int: sa_types.Integer,
            list: sa_types.PickleType,
            str: sa_types.String,
        }

        for field_name, field_type in sample_definition.items():
            # Map the field if not in attributes
            if field_name not in cls.__dict__:
                type_args = []
                type_kwargs = {}

                # Get the SA type
                sa_type_cls = field_mapping.get(field_type)

                # Default to the text type if no mapping is found
                if not sa_type_cls:
                    sa_type_cls = sa_types.String

                # Build the column arguments
                if field_name == "id":
                    col_args = {
                        "primary_key": True,
                        "nullable": False,
                        "unique": True,
                    }
                else:
                    col_args = {}

                # Update the attributes of the class
                setattr(
                    cls,
                    field_name,
                    Column(sa_type_cls(*type_args, **type_kwargs), **col_args),
                )
        super().__init__(classname, bases, dict_)


class BaseModel:
    """This is a Model representing a data schema in the persistence store. A concrete implementation of this
    model has to be provided by each persistence store plugin.
    """
    def __new__(cls, *args, **kwargs):
        if cls is BaseModel:
            raise TypeError("BaseModel cannot be instantiated")
        return super().__new__(cls)

    @classmethod
    @abstractmethod
    def from_entity(cls, entity):
        """Initialize Model object from Entity object"""

    @classmethod
    @abstractmethod
    def to_entity(cls, *args, **kwargs):
        """Convert Model Object to Entity Object"""


@as_declarative(metaclass=DeclarativeMeta)
class SqlalchemyModel(BaseModel):
    """Model representation for the Sqlalchemy Database """

    @declared_attr
    def __tablename__(cls):
        """Snake-case version of the class name"""
        return re.sub(r'(?<!^)(?=[A-Z])', '_', cls.__name__).lower()

    @classmethod
    def from_entity(cls, entity):
        """Initialize Model object from Entity object"""
        pass

    @classmethod
    def to_entity(cls, *args, **kwargs):
        """Convert Model Object to Entity Object"""
        pass


model_cls = type("CustomModel", (SqlalchemyModel,), {})
print(f"Model Cls: {model_cls}")
```

**Error**

With version <1.4.11, all is fine and dandy:

```text
Model Cls: <class 'abc.CustomModel'>
```

With version 1.4.11, an `ArgumentError` is thrown:

```
Traceback (most recent call last):
  File ".../wspace/.../tester_sqlalchemy_cls.py", line 104, in <module>
    model_cls = type("CustomModel", (SqlalchemyModel,), {})
  File ".../wspace/.../tester_sqlalchemy_cls.py", line 62, in __init__
    super().__init__(classname, bases, dict_)
  File ".../.pyenv/versions/python-3.9.4/lib/python3.9/site-packages/sqlalchemy/orm/decl_api.py", line 75, in __init__
    _as_declarative(reg, cls, dict_)
  File ".../.pyenv/versions/python-3.9.4/lib/python3.9/site-packages/sqlalchemy/orm/decl_base.py", line 126, in _as_declarative
    return _MapperConfig.setup_mapping(registry, cls, dict_, None, {})
  File ".../.pyenv/versions/python-3.9.4/lib/python3.9/site-packages/sqlalchemy/orm/decl_base.py", line 177, in setup_mapping
    return cfg_cls(registry, cls_, dict_, table, mapper_kw)
  File ".../.pyenv/versions/python-3.9.4/lib/python3.9/site-packages/sqlalchemy/orm/decl_base.py", line 314, in __init__
    self._early_mapping(mapper_kw)
  File ".../.pyenv/versions/python-3.9.4/lib/python3.9/site-packages/sqlalchemy/orm/decl_base.py", line 200, in _early_mapping
    self.map(mapper_kw)
  File ".../.pyenv/versions/python-3.9.4/lib/python3.9/site-packages/sqlalchemy/orm/decl_base.py", line 979, in map
    mapper_cls(self.cls, self.local_table, **self.mapper_args),
  File "<string>", line 2, in __init__
  File ".../.pyenv/versions/python-3.9.4/lib/python3.9/site-packages/sqlalchemy/util/deprecations.py", line 298, in warned
    return fn(*args, **kwargs)
  File ".../.pyenv/versions/python-3.9.4/lib/python3.9/site-packages/sqlalchemy/orm/mapper.py", line 686, in __init__
    self._configure_pks()
  File ".../.pyenv/versions/python-3.9.4/lib/python3.9/site-packages/sqlalchemy/orm/mapper.py", line 1328, in _configure_pks
    raise sa_exc.ArgumentError(
sqlalchemy.exc.ArgumentError: Mapper mapped class CustomModel->custom_model could not assemble any primary key columns for mapped table 'custom_model'
```

**Versions.**
 - Python: 3.7+
 - SQLAlchemy: 1.4.11

**Additional context**
This is a minimal, reproducible example. The actual code uses a full-blow data container class instead of the dictionary used above (`sample_definition`).

I am not 100% sure, but [the change in this line](https://github.com/sqlalchemy/sqlalchemy/commit/ed3f2c617239668d74ad3d86aeda0ca2030a5933#diff-fb37e8bc18ebc4c14f6780e17648d92847ccbfd5fe96567b76d686f2d7ae210dR570) may be the root of the problem.

I may be constructing the derived `SqlalchemyModel` incorrectly. Please let me know what I should correct in that case.

**Comment by zzzeek** [comment]
hey there -

so there is definitely a behavioral change here between 1.3 and 1.4, however I don't know that it's a bug, at the moment I think 1.4's behavior is more correct, as I made the change intentionally.  In 1.3, the `DeclarativeMeta.__init__` method looks like this:

```py
class DeclarativeMeta(type):
    def __init__(cls, classname, bases, dict_, **kw):
        if "_decl_class_registry" not in cls.__dict__:
            _as_declarative(cls, classname, cls.__dict__)
        type.__init__(cls, classname, bases, dict_)
```

in 1.4, it essentially looks like this:

```py
class DeclarativeMeta(type):
    def __init__(cls, classname, bases, dict_, **kw):
        # ...
        if not cls.__dict__.get("__abstract__", False):
            _as_declarative(reg, cls, dict_)
        type.__init__(cls, classname, bases, dict_)

the difference is 1.3 is using `cls.__dict__` as the "namespace", whereas 1.4 it uses (we have assumed more correctly) the given namespace dictionary of the class, which is the `dict_` argument above.    

The change you refer towards in #6331 had the effect of revealing this difference, whereas previously there was a step that copied class attributes into `dict_`, however this was interfering with some use cases; 1.3 did **not** have this step.   so I dont think #6331 is the issue.  the issue is we are using the namespace dictionary now, and your custom metaclass wishes to generate this namespace, however it's doing it by populating `cls.__dict__` and not `dict_`.       

I think it's more appropriate that since your metaclass is attempting to produce a new namespace for a new class, it should be populating the dictionary, and only optionally the actual `cls.__dict__`:

```py
class DeclarativeMeta(sa_dec.DeclarativeMeta, ABCMeta):

    def __init__(cls, classname, bases, dict_):
        # ...

        for field_name, field_type in sample_definition.items():
            # ...
            # Map the field if not in attributes
            if field_name not in cls.__dict__:

                # ...
                dict_[field_name] = Column(sa_type_cls(*type_args, **type_kwargs), **col_args)

        super().__init__(classname, bases, dict_)
```

all of that said, I've googled around and can't really find clear guidance / background on what a metaclass constructor expects to see in `cls.__dict__` vs the namespace dictionary, and I've always found it confusing that there is overlap in these two dictionaries.    if we can find authoritative takes / docs / background on which dictionary is more "correct", we will look into going with that.   however when I last worked on this, the namespace dictionary seemed to be the more appropriate place to see what attributes have been declared in **this class body**; as your metaclass is attempting to manufacture that, it seems like this is where it should put that information.

**Comment by subhashb** [comment]
Thank you for the quick and in-depth explanation, @zzzeek 🙏 

I did the change you recommended and confirm that everything works 👍 

Amen to the fact that there are no definitive docs that choose between the namespace dictionary vs. the class's `__dict__`. I just went with one instead of the other. I don't see any issues going with the namespace dict in this case, and the change is largely transparent, so all good.
